### PR TITLE
build-fhs-userenv: change to using bubblewrap over chrootenv

### PR DESCRIPTION
**Pull request by illegalprime**
###### Motivation for this change

this makes FHS envs use bubblewrap over Nix's custom `chrootenv.c` which might fix issues such as #33106 but more generally gives us the ability to add many new configuration options with bwrap's added power. Maybe using bwrap will be more stable as well since it seems widely used.

I have tested this by building Yocto images by running bitbake in an FHSUserEnv but I have yet to run steam which is likely the most complicated package that utilizes this.

This comes with fixes from a lot of experience using FHS env for Yocto and attempting to run it in various environments. Namely:

1. `/host/etc` is now `/host-etc` because systemd symlinks `/etc/resolv.conf` to `../run/systemd/resolve/resolv.conf` so keeping `/host-etc` on the same level as `/etc` helps with that.
2. `/etc/pki` is now included so fedora users can get SSL within the env
3. Removing the mount of `/host-etc` (`/host/etc`) and directly bind-mounting all etc files and dirs, then allowing directories of `etc` from the generated rootfs to override the one from the host.

Applications that work:

- [x] `applications/blockchains/mist`
- [x] `applications/editors/android-studio`
- [x] `applications/science/electronics/bitscope`
- [ ] `applications/misc/houdini`
- [x] `applications/misc/lutris`
- [x] `applications/misc/sidequest`
- [ ] `applications/networking/dropbox`
- [ ] `applications/video/lightworks`
- [x] `development/arduino/platformio`
- [x] `games/steam`
- [x] `servers/plex`
- [ ] `build-support/appimage`
- [ ] `tools/package-management/appimage-run`
- [ ] `tools/package-management/conda`

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/build-fhs-userenv/default.nix [hunk 1]
```diff
@@ -1,35 +1,80 @@
...
+  ]);
 
   chrootenv = callPackage ./chrootenv {};
 
-  init = run: writeScript "${name}-init" ''
-    #! ${stdenv.shell}
+  init = run: writeShellScriptBin "${name}-init" ''
     for i in ${env}/* /host/*; do
       path="/''${i##*/}"
       [ -e "$path" ] || ${coreutils}/bin/ln -s "$i" "$path"
```
Comment (hedning) [inline pkgs/build-support/build-fhs-userenv/default.nix]: This can be removed I think, It should be handled by `ro_mounts` and `--ro-bind /etc /host/etc` right?

**Comment by illegalprime** [comment]
The build error is from this line:
https://github.com/NixOS/nixpkgs/pull/55973/files#diff-0e0af17bff7614ce6822d64c6863bb1fR36
but I'm not sure what the fix is, it seems that I can't `readDir "${fhs-env}"` during build because that path doesn't exist yet? it's weird because this works on my test fhs-env. Do i do it at runtime or can i force this directory to exist?

**Comment by hedning** [comment]
ofborg only evaluates things and it's not able to evaluate the buildFHSUserEnv without building due to `readDir`. The easiest fix might be calculate `mounts`, `ro_mounts` and `blacklist` in `bwrap_cmd` instead.

**Comment by illegalprime** [comment]
did some more tests, most things didn't run because my host system's `/etc/fonts/fonts.conf` could not be read by these packages for some reason. Nothing that seems to be wrong with the FHS env itself.
also updated the docs.

**Comment by matthewbauer** [comment]
@grahamcofborg build houdini lightworks conda

**Comment by illegalprime** [comment]
@hedning what do I need to merge this?

**Comment by hedning** [comment]
It would be good to get an OK from @yegortimoshenko or @abbradar first.

File: doc/functions/fhs-environments.xml [hunk 2]
```diff
@@ -50,6 +50,58 @@
...
+      </para>
+    </listitem>
+  </varlistentry>
+  <varlistentry>
+    <term>
+      <literal>bwrap_args</literal>
+    </term>
+    <listitem>
+      <para>
+        Any extra arguments tp pass to bubblewrap (which is now the underlying
```
Comment (hedning) [inline doc/functions/fhs-environments.xml]: typo: tp -> to

File: doc/functions/fhs-environments.xml [hunk 3]
```diff
@@ -50,6 +50,58 @@
     </para>
    </listitem>
   </varlistentry>
+  <varlistentry>
+    <term>
+      <literal>automount</literal>
```
Comment (yegortimoshenko) [inline doc/functions/fhs-environments.xml]: I don't think this should be an option, not for `buildFHSUserEnv`, that is: it should really be the only behavior, because isolation is not the goal (FHS compat is).

Later, we can get to sandboxing of individual derivations, XDG portals, etc. That said, it will certainly require a complete redesign of `buildFHSUserEnv` and will have to be carefully thought out.

**Review by lukateras — CHANGES_REQUESTED** [review]
I would prefer if `bwrap` replacement were to replace `chrootenv` without extending API.

File: doc/functions/fhs-environments.xml [hunk 4]
```diff
@@ -50,6 +50,58 @@
...
+      </para>
+    </listitem>
+  </varlistentry>
+  <varlistentry>
+    <term>
+      <literal>chdir</literal>
+    </term>
+    <listitem>
+      <para>
+        Set to <literal>"$(pwd)"</literal> by default, this option determines
```
Comment (yegortimoshenko) [inline doc/functions/fhs-environments.xml]: This seems to be orthogonal to FHS emulation (and can be/is implemented in wrappers).

**Comment by illegalprime** [comment]
@yegortimoshenko if I wanted to edit the bwrap options for my own uses, how should I implement this to make that possible? Maybe I should include it as an arg in the derivation and use `overrideAttrs`? Or maybe I should just make my own custom `FHSUserEnv`, but I want to re-use as much upstream code as I can.

**Comment by yegortimoshenko** [comment]
It's just that I worry that we might need to switch back at some point, if we ever need anything bubblewrap doesn't implement. If we were to expose bubblewrap options, we will never be able to change underlying implementation without breaking reverse compat.

**Comment by abbradar** [comment]
Yeah, that's what I fear too so my current idea is to proceed with caution and 1) don't expose the implementation 2)  look at bubblewrap's code to see if it's hackable enough. 

On May 28, 2019 12:13:58 PM GMT+03:00, Yegor Timoshenko <notifications@github.com> wrote:
>It's just that I worry that we might need to switch back at some point,
>if we ever need anything bubblewrap doesn't implement. If we were to
>expose bubblewrap options, we will never be able to change underlying
>implementation without breaking reverse compat.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/55973#issuecomment-496433818

-- 
Nikolay.

**Comment by abbradar** [comment]
I didn't yet review the PR itself yet, only bubblewrap -- it's quite more complicated but we should definitely try using it; the code is readable too.

One concern of mine is if nested bubblewrap would work okay, especially if it would blow `/proc/mounts`. This needs to be tested.

Another thing -- we explicitly _don't_ want any kind of isolation that bubblewrap does, including no automount, any other namespaces etc, and we also don't want to expose it as an implementation. I think we could hit both birds by adding `buildBwrapUserEnv` upon which old `buildFHSUserEnv` resides. The former would have all the bubblewrap options exposed and the latter would enforce the old behaviour.

**Comment by yegortimoshenko** [comment]
Wrt naming: `User` seems to be redundant, maybe `buildBwrapEnv`?

**Comment by abbradar** [comment]
Maybe; don't have big preferences here. @illegalprime can choose the naming ~_^

**Comment by vcunat** [comment]
For reference, the `relocation error` messages from above [are claimed](https://sourceware.org/ml/libc-alpha/2019-08/msg00029.html) to be gone since glibc 2.30, which has [a WIP PR](https://github.com/NixOS/nixpkgs/pull/66528).

**Comment by illegalprime** [comment]
I'm back!
@yegortimoshenko I updated this branch to not introduce any new options and I also changed `/host/etc` to `/host-etc` because `/etc/resolv.conf` is symlinked to `../run/` in systemd. Fedora also stores its SSL certificates as a symlink to somewhere not included in the current `etc` package so that should be investigated too.

**Comment by illegalprime** [comment]
OK Fedora stores its SSL certs in `/etc/pki` so I've added that to the environment.

**Comment by illegalprime** [comment]
I added a commit which removes `/host/etc` (or `/host-etc`) and bind-mounts specific files or directories from the host's `/etc` to the chroot. It then will look inside the generated root filesystem's `/etc` and bind mount subdirectories of it. because this is all subdirectories of etc the generated etc and the host etc are "merged" somewhat at runtime.

This helps avoid weird symlink bugs when nesting chroots like in this commit 06f27dc2e932778cf1d2ce6b0bf30b32575dd16a and in general simplifies things (IMO)

**Comment by zimbatm** [comment]
:+1: really cool work. Do you need help to finish this PR?

**Comment by bachp** [comment]
@illegalprime Are you still working on this. Looks very promising

**Comment by Atemu** [comment]
Rebased on top of nixos-unstable here: https://github.com/Atemu/nixpkgs/tree/chrootenv2bubblewrap

File: pkgs/build-support/build-fhs-userenv/default.nix [hunk 5]
```diff
@@ -1,35 +1,113 @@
...
+    done
+
+    exec ${bubblewrap}/bin/bwrap \
+      --dev-bind /dev /dev \
+      --proc /proc \
+      --chdir "$(pwd)" \
+      --unshare-all \
+      --share-net \
+      --die-with-parent \
+      --ro-bind /nix /nix \
```
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv/default.nix:97]: Would this break single user nix?

**Comment by Atemu** [comment]
Since the compatibility is not 100% yet, perhaps we should limit bubblewrap use to the programs which are known to work first and move over the rest in later PRs.

Making all of these work feels like too large of a scope to me.

**Comment by Mic92** [comment]
We could call it v2 and migrate everything to it and at some point delete the old version.

**Comment by Atemu** [comment]
I did some rebase magic to do just that in a transparent manner https://github.com/Atemu/nixpkgs/tree/buildFHSUserEnvBw

Should I open a new PR?

Once we've ported all packages over, we can simply rename the function to the old name. This name should be temporary.

**Comment by Mic92** [comment]
> I did some rebase magic to do just that in a transparent manner [Atemu/nixpkgs@`buildFHSUserEnvBw`](https://github.com/Atemu/nixpkgs/tree/buildFHSUserEnvBw?rgh-link-date=2020-07-31T13%3A26%3A57Z)
> I also ported three of the FHS packages that I know work fine.
> 
> Should I open a new PR?
> 
> Once we've ported all packages over, we can simply rename the function to the old name. This name should be temporary.

bring it on.

**Comment by xaverdh** [comment]
With #94442 merged, should this be closed now?

**Comment by siraben** [comment]
#94442 seems to subsume this PR, closing.
